File: app/src/main/java/fr/free/nrw/commons/upload/SpinnerLanguagesAdapter.java
package fr.free.nrw.commons.upload;

import android.content.Context;
import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import fr.free.nrw.commons.R;
import fr.free.nrw.commons.utils.BiMap;
import fr.free.nrw.commons.utils.LangCodeUtils;


public class SpinnerLanguagesAdapter extends ArrayAdapter {

    private final int resource;
    private final LayoutInflater layoutInflater;
    private List<String> languageNamesList;
    private List<String> languageCodesList;
    private final BiMap<AdapterView, String> selectedLanguages;
    public String selectedLangCode="";
    private Context context;
    private boolean dropDownClicked;
    private String savedLanguageValue;



    public SpinnerLanguagesAdapter(@NonNull Context context,
                                   int resource,
                                   BiMap<AdapterView, String> selectedLanguages,
                                   String savedLanguageValue) {
        super(context, resource);
        this.resource = resource;
        this.layoutInflater = LayoutInflater.from(context);
        languageNamesList = new ArrayList<>();
        languageCodesList = new ArrayList<>();
        prepareLanguages();
        this.selectedLanguages = selectedLanguages;
        this.context = context;
        this.dropDownClicked = false;
        this.savedLanguageValue = savedLanguageValue;
    }

    private void prepareLanguages() {
        List<Language> languages = getLocaleSupportedByDevice();

        for(Language language: languages) {
            if(!languageCodesList.contains(language.getLocale().getLanguage())) {
                languageNamesList.add(language.getLocale().getDisplayName());
                languageCodesList.add(language.getLocale().getLanguage());
            }
        }
    }

    private List<Language> getLocaleSupportedByDevice() {
        List<Language> languages = new ArrayList<>();
        Locale[] localesArray = Locale.getAvailableLocales();
        for (Locale locale : localesArray) {
            languages.add(new Language(locale));
        }

        Collections.sort(languages, (language, t1) -> language.getLocale().getDisplayName()
                .compareTo(t1.getLocale().getDisplayName()));
        return languages;
    }

    @Override
    public boolean isEnabled(int position) {
        return !languageCodesList.get(position).isEmpty()&&
                (!selectedLanguages.containsKey(languageCodesList.get(position)) ||
                        languageCodesList.get(position).equals(selectedLangCode));
    }

    @Override
    public int getCount() {
        return languageNamesList.size();
    }

    @Override
    public View getDropDownView(int position, @Nullable View convertView,
                                @NonNull ViewGroup parent) {
        if (convertView == null) {
            convertView = layoutInflater.inflate(resource, parent, false);
        }
        ViewHolder holder = new ViewHolder(convertView);
        holder.init(position, true, savedLanguageValue);

        dropDownClicked = true;
        return convertView;
    }

    @Override
    public @NonNull
    View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        ViewHolder holder;
        if (convertView == null) {
            convertView = layoutInflater.inflate(resource, parent, false);
            holder = new ViewHolder(convertView);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }
        holder.init(position, false, savedLanguageValue);
        return convertView;
    }


    public class ViewHolder {

        @BindView(R.id.tv_language)
        TextView tvLanguage;

        @BindView(R.id.view)
        View view;

        public ViewHolder(View itemView) {
            ButterKnife.bind(this, itemView);
        }

        public void init(int position, boolean isDropDownView, String savedLanguageValue) {
            String languageCode = LangCodeUtils.fixLanguageCode(languageCodesList.get(position));
            final String languageName = StringUtils.capitalize(languageNamesList.get(position));

            if (!isDropDownView) {
                view.setVisibility(View.GONE);
                if (languageCode.length() > 2)
                    tvLanguage.setText(languageCode.substring(0, 2));
                else
                    tvLanguage.setText(languageCode);
            } else {
                view.setVisibility(View.VISIBLE);
                if (languageCodesList.get(position).isEmpty()) {
                    tvLanguage.setText(languageName);
                    tvLanguage.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
                } else {
                    tvLanguage.setText(
                            String.format("%s [%s]", languageName, languageCode));
                    if (selectedLanguages.containsKey(languageCodesList.get(position)) &&
                            !languageCodesList.get(position).equals(selectedLangCode)) {
                        tvLanguage.setTextColor(Color.GRAY);
                    }
                }
            }
        }
    }

    String getLanguageCode(int position) {
        return languageCodesList.get(position);
    }

    int getIndexOfUserDefaultLocale(Context context) {
        return languageCodesList.indexOf(context.getResources().getConfiguration().locale.getLanguage());
    }

    int getIndexOfLanguageCode(String languageCode) {
        return languageCodesList.indexOf(languageCode);
    }
}

File: app/src/main/java/fr/free/nrw/commons/upload/SpinnerLanguagesAdapter.kt
package fr.free.nrw.commons.upload

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import androidx.annotation.LayoutRes
import androidx.core.os.ConfigurationCompat
import fr.free.nrw.commons.R
import fr.free.nrw.commons.utils.LangCodeUtils
import kotlinx.android.synthetic.main.row_item_languages_spinner.*
import kotlinx.android.synthetic.main.row_item_languages_spinner.view.*
import org.apache.commons.lang3.StringUtils
import org.wikipedia.language.AppLanguageLookUpTable
import java.util.*

/**
 * This class handles the display of language spinners and their dropdown views for UploadMediaDetailFragment
 *
 * @property selectedLanguages - controls the enabled state of dropdown views
 *
 * @param context - required by super constructor
 */
class SpinnerLanguagesAdapter constructor(
    context: Context,
    private val selectedLanguages: HashMap<*, String>,
    private var languageNamesList: List<String>,
    private val languageCodesList: List<String>
) : ArrayAdapter<String?>(context, R.layout.row_item_languages_spinner) {

    var language: AppLanguageLookUpTable = AppLanguageLookUpTable(context)
    private val filter = LanguageFilter()
    var selectedLangCode = ""

    override fun isEnabled(position: Int) = languageCodesList[position].let {
        it.isNotEmpty() && !selectedLanguages.containsValue(it) && it != selectedLangCode
    }

    override fun getCount() = languageNamesList.size


    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val rowView = parent.inflate(R.layout.row_item_languages_spinner)
        val languageCode = languageCodesList[position]
        val languageName = languageNamesList[position]
        rowView.tv_language.let {
            it.isEnabled = isEnabled(position)
            if (languageCode.isEmpty()) {
                it.text = StringUtils.capitalize(languageName)
                it.textAlignment = View.TEXT_ALIGNMENT_CENTER
            } else {
                it.text =
                    "${StringUtils.capitalize(languageName)}" +
                            " [${LangCodeUtils.fixLanguageCode(languageCode)}]"
            }
        }
        return rowView
    }

    fun getLanguageCode(position: Int): String {
        return languageCodesList[position]
    }

    fun getIndexOfUserDefaultLocale(context: Context): Int {
        return languageCodesList.indexOf(context.locale.language)
    }

    fun getIndexOfLanguageCode(languageCode: String): Int {
        return languageCodesList.indexOf(languageCode)
    }


    override fun getFilter() = filter

    inner class LanguageFilter() : Filter() {

        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val filterResults = FilterResults()
            val tempList: ArrayList<String> = ArrayList<String>()
            if (constraint != null && language.localizedNames != null) {
                val length: Int = language.localizedNames.size
                var i = 0
                while (i < length) {
                    val item: String = language.localizedNames.get(i)
                    val uilang = getIndexOfUserDefaultLocale(context)
                    if(item.contains(constraint, true) || Locale(languageCodesList[i]).getDisplayName(
                            Locale(languageCodesList[uilang])).contains(constraint, true))
                        tempList.add(item)
                    i++
                }
                filterResults.values = tempList
                filterResults.count = tempList.size
            }
            return filterResults
        }

        override fun publishResults(constraint: CharSequence?, results: FilterResults) {
            languageNamesList = results.values as List<String>
            if (results.count > 0) {
                notifyDataSetChanged()
            } else {
                notifyDataSetInvalidated()
            }

        }

    }

}

private fun ViewGroup.inflate(@LayoutRes resId: Int) =
    LayoutInflater.from(context).inflate(resId, this, false)

private val Context.locale: Locale
    get() = ConfigurationCompat.getLocales(resources.configuration)[0]
